File: World.cpp
//
// Created by karol on 28.03.2025.
//

#include "include/World.h"
#include "include/Animal.h"

using namespace std;


World::World(int width, int height ) {
    this->width = width;
    this->height = height;
    human = nullptr;
    i = 0;
}

World::~World() {
    organisms.clear();
}

void World::update(char input) {
    sort(organisms.begin(), organisms.end(), [](const Organism* a, const Organism* b) {
        if (a->getInitiative() == b->getInitiative()) {
            return a->getAge() > b->getAge();
        }
        return a->getInitiative() > b->getInitiative();
    });

    for (Organism* organism : organisms) {
        organism->setHasActed(false);
    }

    for (i = 0; i < organisms.size(); i++) {
        organisms[i]->action(input);
    }
}

void World::drawWorld(int width, int height) {
    system("cls");

    drawHorizontalBorder(width);

    for (int i = 0; i < height; i++) {
        std::cout << "#";

        for (int j = 0; j < width; j++) {
            Organism* org = getAtCoordinates(Point(j, i));
            if (org != nullptr)
                std::cout << org->getSymbol();
            else
                std::cout << "  ";
        }

        std::cout << " # " << std::endl;
    }

    drawHorizontalBorder(width);
}

void World::remove(const Point& position) {
    if (getAtCoordinates(position) == human) {
        human = nullptr;
    }
    
    for (auto it = organisms.begin(); it != organisms.end(); it++) {
        if ((*it)->getPosition() == position) {
            if (distance(organisms.begin(), it) <= i) {
                i--;
            }
            organisms.erase(it);
            return;
        }
    }
}

Point World::getRandomNeighbor(const Point& position) const {
    std::vector<vector<int>> displaces = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

    std::random_device rd;
    std::default_random_engine rng(rd());

    std::shuffle(displaces.begin(), displaces.end(), rng);

    for (vector<int> displace : displaces) {
        Point neighbor(position.x + displace[0], position.y + displace[1]);

        if (isWithinBounds(neighbor)) {
            return neighbor; // Zwróć pierwsze wolne sąsiednie pole
        }
    }
    
    return Point(-1, -1);
}

Organism* World::getAtCoordinates(Point cords) const {
    for (Organism* org : organisms) {
        if (org->getPosition() == cords) return org;
    }
    return nullptr;
}

bool World::isWithinBounds(const Point& position) const {
    return position.x >= 0 && position.x < width && position.y >= 0 && position.y < height;
}

bool World::isHumanAlive() const {
    for (const auto& organism : organisms) {
        if (organism == human) {
            return true;
        }
    }
    return false;
}

Point World::getRandomFreeSpace() const {
    for(int i = 0; i < width * height; i++) {
        int x = rand() % width;
        int y = rand() % height;
        Point randomPoint(x, y);

        if(getAtCoordinates(randomPoint) == nullptr) {
            return randomPoint;
        }
    }

    return Point(-1, -1);

}

void World::killNeighbors(Point& position, int type) {
    // type = 1 - animals, type = 0 - everything
    const int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    const int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};

    for (int i = 0; i < 8; i++) {
        Point neighbor(position.x + dx[i], position.y + dy[i]);
        Organism* killer = getAtCoordinates(position);

        if (isWithinBounds(neighbor)) {
            Organism* other = getAtCoordinates(neighbor);
            if (other != nullptr) {
                if (type == 1 && dynamic_cast<Animal*>(other) != nullptr) {
                    killer->kill(*other);
                } else if (type == 0) {
                    killer->kill(*other);
                }
            }
        }
    }
}

Point World::getRandomFreeSpaceAround(const Point& position) const {
    std::vector<int> dx = {0, 1, -1};
    std::vector<int> dy = {0, 1, -1};

    std::random_device rd;
    std::default_random_engine rng(rd());

    std::shuffle(dx.begin(), dx.end(), rng);
    std::shuffle(dy.begin(), dy.end(), rng);

    for (int x : dx) {
        for (int y : dy) {
            Point neighbor(position.x + x, position.y + y);

            if (isWithinBounds(neighbor) && getAtCoordinates(neighbor) == nullptr) {
                return neighbor; // Zwróć pierwsze wolne sąsiednie pole
            }

        }
    }
    
    return Point(-1, -1);
}

Point World::findSafeSpaceAround(Point& position) const {
    std::vector<vector<int>> displaces = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

    std::random_device rd;
    std::default_random_engine rng(rd());

    std::shuffle(displaces.begin(), displaces.end(), rng);

    Organism* org = getAtCoordinates(position);

    for (vector<int> displace : displaces) {
        Point neighbor(position.x + displace[0], position.y + displace[1]);

        Organism* other = getAtCoordinates(neighbor);
        if (other != nullptr && other->getStrength() > org->getStrength()) {
            continue; // Skip if the neighbor is stronger
        }

        if (isWithinBounds(neighbor)) {
            return neighbor;
        }
    }
    return position;
}

bool World::hasFreeSpaceAround(Point position) {
    const int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};
    const int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};

    for (int i = 0; i < 8; i++) {
        Point neighbor(position.x + dx[i], position.y + dy[i]);

        if (isWithinBounds(neighbor) && getAtCoordinates(neighbor) == nullptr) {
            return true;
        }
    }

    return false;
}

Point World::getRandomSpaceDoubleMove(const Point& position) const {
    std::vector<std::pair<int, int>> displacements = {
        {0, 1}, {0, -1}, {1, 0}, {-1, 0}, // Single moves
        {0, 2}, {0, -2}, {2, 0}, {-2, 0}  // Double moves
    };

    std::random_device rd;
    std::default_random_engine rng(rd());

    std::shuffle(displacements.begin(), displacements.end(), rng);

    for (const auto& displacement : displacements) {
        Point neighbor(position.x + displacement.first, position.y + displacement.second);

        if (isWithinBounds(neighbor)) {
            return neighbor;
        }
    }

    return position;
}

void World::move(const Point& position, const Point &destination) {
    Organism* org = getAtCoordinates(position);
    if (org != nullptr && isWithinBounds(destination) && getAtCoordinates(destination) == nullptr) {
        org->setPosition(destination);
    }
}

void World::drawHorizontalBorder(int width) {
    for (int i = 0; i < width + 2; i++) {
        cout << "# ";
    }
    cout << endl;
}

void World::spawnOrganism(Organism* organism) {
    organisms.push_back(organism);
    organism->setPosition(Point(rand() % width, rand() % height));
}


void World::spawnOrganism(Organism* organism, const Point& position) {
    organisms.push_back(organism);
    organism->setPosition(position);
}

void World::addShoutSummaryMessage(const std::string& message) {
    shoutSummaryMessages.push_back(message);
}

void World::printShoutSummaryMessages() {
    int messageCount = shoutSummaryMessages.size() > 10 ? 10 : shoutSummaryMessages.size();

    for(int i = 0; i < messageCount; i++) {
        std::cout << shoutSummaryMessages[i] << std::endl;
    }
    shoutSummaryMessages.clear();
}

void World::printShoutSummary() {
    std::cout << std::endl;
    std::cout << "Shout actions summary:" << std::endl;
    std::cout << "----------------" << std::endl;
    
    printShoutSummaryMessages();
}

void World::printHumanInfo() {
    if (human != nullptr) {
        std::cout << "Human - position: (" << human->getPosition().x <<
        ", " << human->getPosition().y <<
        ") strength: " << human->getStrength() <<  std::endl;
    }
}

Point World::getPlayerPosition() const {
    return human->getPosition();
}

int World::getHeight() {
    return height;
}

int World::getWidth() {
    return width;
}

void World::setHuman(Organism* org) {
    human = org;
}

Organism* World::getHuman() const {
    return human;
}
std::vector<Organism*>& World::getOrganisms() {
    return organisms;
}

File: game-of-life-java/src/World.java
import lib.Size;

import java.awt.*;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Comparator;
import java.io.Serializable;

public class World implements Serializable {
    private static final long serialVersionUID = 1L;

    private List<Organism> organisms;
    private Size size;
    private transient Window window;

    public void setWindow(Window window) {
        this.window = window;
    }

    public void setWorldReferenceForAll() {
        for (Organism org : organisms) {
            org.setWorld(this);
        }
    }

    public World(Size size) {
        this.organisms = new ArrayList<Organism>();
        this.size = size;
    }

    public Size getSize() {
        return size;
    }

    public List<Organism> getOrganisms() {
        return organisms;
    }

    public Window getWindow() {
        return window;
    }

    public void update(String input) {
        // Sortowanie organizmów według inicjatywy, a następnie wieku
        organisms.sort(Comparator.comparingInt(Organism::getInitiative).reversed()
                                  .thenComparingInt(Organism::getAge));

        for(Organism organism : organisms) {
            organism.setHasActed(false);
        }

        List<Organism> copy = new ArrayList<>(organisms);
        for (Organism organism : copy) {
            if (!organisms.contains(organism)) continue; // skip if killed this turn
            organism.action(input);
        }

        System.out.println("Number of organisms: " + organisms.size());
    }

    public boolean isTileOccupied(Point position) {
        for (Organism organism : organisms) {
            if (organism.getPosition().equals(position)) {
                return true; // Kafelek jest zajęty
            }
        }
        return false; // Kafelek jest wolny
    }

    public void addOrganism(Organism organism) {
        if (!isTileOccupied(organism.getPosition())) {
            organisms.add(organism);
        } else {
            System.out.println("Tile at " + organism.getPosition() + " is already occupied!");
        }
    }

    public void removeOrganism(Organism organism) {
        organisms.remove(organism);
    }

    public boolean isWithinBounds(Point position) {
        return position.x >= 0 && position.x < size.x && position.y >= 0 && position.y < size.y;
    }

    public Point generateRandomPosition(Point currentPosition) {
        int[][] directions = { {0, -1}, {0, 1}, {1, 0}, {-1, 0} };
        List<Point> validPositions = new ArrayList<>();

        for (int[] d : directions) {
            Point newPos = new Point(currentPosition.x + d[0], currentPosition.y + d[1]);
            if (isWithinBounds(newPos)) {
                validPositions.add(newPos);
            }
        }

        if (!validPositions.isEmpty()) {
            return validPositions.get((int) (Math.random() * validPositions.size()));
        }
        return null; // Brak dostępnych pozycji
    }

    // Nowa metoda generująca losową pozycję w obrębie planszy
    public Point generateRandomPosition() {
        java.util.Random random = new java.util.Random();
        Point position;
        do {
            int x = random.nextInt(size.x);
            int y = random.nextInt(size.y);
            position = new Point(x, y);
        } while (isTileOccupied(position)); // Sprawdzanie, czy pole nie jest zajęte
        return position;
    }

    public Point generateSafePosition(Point position) {
        List<int[]> displacements = new ArrayList<>(java.util.Arrays.asList(
                new int[]{0, 1}, new int[]{1, 0}, new int[]{-1, 0}, new int[]{0, -1}
        ));

        Collections.shuffle(displacements); // Losowe przetasowanie kierunków

        Organism org = getOrganismAtPosition(position);

        for (int[] displacement : displacements) {
            Point neighbor = new Point(position.x + displacement[0], position.y + displacement[1]);

            if (isWithinBounds(neighbor)) { // Sprawdzanie granic planszy
                Organism other = getOrganismAtPosition(neighbor);
                if (other != null && other.getStrength() > org.getStrength()) {
                    continue; // Pomijamy, jeśli sąsiad jest silniejszy
                }
                return neighbor; // Zwracamy bezpieczną pozycję
            }
        }

        return position; // Jeśli brak bezpiecznych pozycji, zwracamy bieżącą pozycję
    }

    public Organism getOrganismAtPosition(Point position) {
        for (Organism organism : organisms) {
            if (organism.getPosition().equals(position)) {
                return organism;
            }
        }
        return null; // Brak organizmu na tej pozycji
    }

    public void createOrganism(Class<? extends Organism> organismClass, int count, int age) {
        for (int i = 0; i < count; i++) {
            Point position = generateRandomPosition();
            Organism organism = null;
            try {
                organism = organismClass.getConstructor(Point.class, World.class, int.class)
                        .newInstance(position, this, age);
            } catch (InvocationTargetException | NoSuchMethodException | InstantiationException | IllegalAccessException e) {
                throw new RuntimeException(e);
            }
            addOrganism(organism);
        }
    }

    public void createOrganismAtPosition(Class<? extends Organism> organismClass, Point position, int age) {
        Organism organism = null;
        try {
            organism = organismClass.getConstructor(Point.class, World.class, int.class)
                    .newInstance(position, this, age);
        } catch (InstantiationException | NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
        addOrganism(organism);
    }

    public void killNeighbors(Point position, int type) {
        // type = 1 - animals, type = 0 - everything
        final int[] dx = {-1, -1, -1, 0, 0, 1, 1, 1};
        final int[] dy = {-1, 0, 1, -1, 1, -1, 0, 1};

        for (int i = 0; i < 8; i++) {
            Point neighbor = new Point(position.x + dx[i], position.y + dy[i]);
            if (isWithinBounds(neighbor)) { // Sprawdzanie granic planszy
                Organism other = getOrganismAtPosition(neighbor);
                if (other != null) {
                    if (type == 1 && other instanceof Animal) {
                        removeOrganism(other);
                        String message = "Organism " + other.getClass().getSimpleName() + " was killed by Hogweed";
                        window.addMessage(message);
                    } else if (type == 0) {
                        removeOrganism(other);
                        String message = "Organism " + other.getClass().getSimpleName() + " was killed by Human";
                        window.addMessage(message);
                    }
                }
            }
        }
    }

    public Point getRandomSpaceDoubleMove(Point position) {
        List<int[]> displacements = new ArrayList<>(java.util.Arrays.asList(
            new int[]{0, 1}, new int[]{0, -1}, new int[]{1, 0}, new int[]{-1, 0}, // Single moves
            new int[]{0, 2}, new int[]{0, -2}, new int[]{2, 0}, new int[]{-2, 0}  // Double moves
        ));

        java.util.Collections.shuffle(displacements); // Losowe przetasowanie kierunków

        for (int[] displacement : displacements) {
            Point neighbor = new Point(position.x + displacement[0], position.y + displacement[1]);

            if (isWithinBounds(neighbor)) { // Sprawdzanie granic planszy
                return neighbor; // Zwracamy pierwszą dostępną pozycję
            }
        }

        return position; // Jeśli brak dostępnych pozycji, zwracamy bieżącą pozycję
    }

    public Point getRandomFreeSpaceAround(Point position) {
        List<int[]> displacements = new ArrayList<>(java.util.Arrays.asList(
                new int[]{0, 1}, new int[]{0, -1}, new int[]{1, 0}, new int[]{-1, 0}, // Single moves
                new int[]{1, 1}, new int[]{1, -1}, new int[]{-1, 1}, new int[]{-1, -1} // Diagonal moves
        ));

        java.util.Collections.shuffle(displacements); // Losowe przetasowanie kierunków

        for (int[] displacement : displacements) {
            Point neighbor = new Point(position.x + displacement[0], position.y + displacement[1]);

            if (isWithinBounds(neighbor) && getOrganismAtPosition(neighbor) == null) {
                return neighbor; // Zwracamy pierwsze wolne sąsiednie pole
            }
        }

        return new Point(-1, -1); // Jeśli brak wolnych pól, zwracamy nieprawidłową pozycję
    }

    public boolean hasFreeSpaceAround(Point position) {
        final int[] dx = {-1, -1, -1, 0, 0, 1, 1, 1};
        final int[] dy = {-1, 0, 1, -1, 1, -1, 0, 1};

        for (int i = 0; i < 8; i++) {
            Point neighbor = new Point(position.x + dx[i], position.y + dy[i]);

            if (isWithinBounds(neighbor) && getOrganismAtPosition(neighbor) == null) {
                return true; // Znaleziono wolne pole
            }
        }
        return false; // Brak wolnych pól
    }
}